### PR TITLE
Chapter 10: Secrets

### PR DESCRIPTION
**Pull request by EdwinVW**
Chapter 10 - Secrets

### CODE REVIEW

**Comment by EdwinVW** [comment]
This is chapter 10 about the Secrets Management building block.

@nishanil / @IEvangelist please review. Thanks!

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+- Azure Key Vault
+- GCP Secret Manager
+- HashiCorp Vault
+
+ > [!IMPORTANT]
+ > The environment variables and local file components are designed for development workloads only.
+
+The following sections show how to configure a secrets management component.
+
+### Configuration (TODO Move generic text to getting started)
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:166]:
```suggestion
### Configuration
```

Not sure if this TODO was done or not?

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+> The path to a secrets file can be a absolute or relative path. The relative path is based on the folder in which the application starts. In the example, the `components` folder is a sub-folder of the directory that contains the .NET application.
+
+From the application folder, start the Dapr application specifying the `components` path as a command-line argument:
+
+```bash
+dapr run --app-id basket-api --components-path ./components dotnet run
+```
+
+ > [!NOTE]
+ > This above example applies to running Dapr in stand-alone mode. For Kubernetes hosting, consider using volume mounts.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:251]:
```suggestion
> [!NOTE]
> This above example applies to running Dapr in stand-alone mode. For Kubernetes hosting, consider using volume mounts.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+        "customerdb": "some connection string",
+        "productdb": "some connection string"
+    }
+}
+```
+
+Using a colon as a separator, you can retrieve the `customerdb` connection-string using the key `connectionStrings:customerdb`.
+
+ > [!NOTE]
+ > The colon is the default separator value.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:268]:
```suggestion
> [!NOTE]
> The colon `:` is the default separator value.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+      name: eshopsecrets
+      key: redisPassword
+auth:
+  secretStore: kubernetes
+```
+
+The `secretKeyRef` element specifies the name of the Kubernetes secret and the secret's key, `eshopsecrets`, and `redisPassword` respectively. The `auth` metadata section instructs Dapr to use the Kubernetes secrets management component.
+
+ > [!NOTE]
+ > Auth is the default value when using Kubernetes secrets and can be omitted.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:327]:
```suggestion
> [!NOTE]
> Auth is the default value when using Kubernetes secrets and can be omitted.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+
+For this example to work, the following prerequisites must be satisfied:
+
+- You've secured administrative access to an Azure subscription.
+- You've provisioned an Azure Key Vault named `eshopkv` that holds a secret named `redisPassword` that contains the password for connecting to the Redis server.
+- You've created [service principal](https://docs.microsoft.com/azure/active-directory/develop/howto-create-service-principal-portal) in Azure Active Directory.
+- You've installed an X509 certificate for this service principal (containing both the public and private key) on the local filesystem.
+
+ > [!NOTE]
+ > A service principal is an identity that can be used by an application to authenticate an Azure service. The service principal uses a X509 certificate. The application uses this certificate as a credential to authenticate itself.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:343]:
```suggestion
> [!NOTE]
> A service principal is an identity that can be used by an application to authenticate an Azure service. The service principal uses a X509 certificate. The application uses this certificate as a credential to authenticate itself.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+
+The secrets management component type is `secretstores.azure.keyvault`. The `metadata` element to configure access to Key Vault requires the following properties:
+
+- The `vaultName` contains the name of the Azure Key Vault.
+- The `spnTenantId` contains the *tenant id* of the service principal used to authenticate against the Key Vault.
+- The `spnClientId` contains the *app id* of the service principal used to authenticate against the Key Vault.
+- The `spnCertificateFile` contains the path to the certificate file for the service principal to authenticate against the Key Vault.
+
+> [!NOTE]
+> You can copy the service principal information from the Azure portal or Azure CLI .
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:379]:
```suggestion
> [!TIP]
> You can copy the service principal information from the Azure portal or Azure CLI .
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+
+Now the application can retrieve the Redis password from the Azure Key Vault.
+
+#### Using Key Vault when running on Kubernetes
+
+Consuming Azure KeyVault with Dapr and Kubernetes also requires a service principal to authenticate against the Azure Key Vault.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:385]:
```suggestion
Consuming Azure Key Vault with Dapr and Kubernetes also requires a service principal to authenticate against the Azure Key Vault.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+
+> [!IMPORTANT]
+> It's critical to keep the X509 certificate file for the service principal in a safe place. It's best to place it in a well-known folder outside the source-code repository. The configuration file can then reference the certificate file from this well-known folder. On a local development machine, you're responsible for copying the certificate to the folder. For automated deployments, the pipeline will copy the certificate to the machine where the application is deployed. It's a best practice to use a different service principal per environment. Doing so prevents the service principal from a DEVELOPMENT environment to access secrets in a PRODUCTION environment.
+
+When running in AKS, it's preferable to use an [Azure managed identity](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview) for authenticating against Azure Key Vault. Managed identities are outside of the scope of this book, but explained in the [Azure KeyVault with managed identities](https://docs.dapr.io/operations/components/setup-secret-store/supported-secret-stores/azure-keyvault-managed-identity/) documentation.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:429]:
```suggestion
When running in  Azure Kubernetes Service (AKS), it's preferable to use an [Azure Managed Identity](https://docs.microsoft.com/azure/active-directory/managed-identities-azure-resources/overview) for authenticating against Azure Key Vault. Managed identities are outside of the scope of this book, but explained in the [Azure Key Vault with managed identities](https://docs.dapr.io/operations/components/setup-secret-store/supported-secret-stores/azure-keyvault-managed-identity/) documentation.
```

The first time we use an acronym we need to fully qualify it.

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+      "-components-path", "/components",
+      "-config", "/configuration/eshop-config.yaml"
+      ]
+    volumes:
+      - "./dapr/components/:/components"
+      - "./dapr/configuration/:/configuration"
+```
+
+ > [!Note]
+ > The Docker Compose override file contains environmental specific configuration values.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:511]:
```suggestion
> [!NOTE]
> The Docker Compose override file contains environmental specific configuration values.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 10]
```diff
@@ -0,0 +1,553 @@
...
+
+Note the `/components` volume mount and `--components-path` command-line argument passed into the `daprd` startup command.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:513]:
```suggestion
The `/components` volume mount and `--components-path` command-line argument are passed into the `daprd` startup command.
```

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md [hunk 11]
```diff
@@ -0,0 +1,553 @@
...
+  - name: redisHost
+    value: redis:6379
+  - name: redisPassword
+    secretKeyRef:
+      name: redisPassword
+auth:
+  secretStore: eshop-secretstore
+```
+
+Note how the example references secrets from the local Redis store.
```
Comment (IEvangelist) [inline docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:535]:
```suggestion
In the preceding example, the local Redis store is used to reference secrets.
```

**Review by IEvangelist — CHANGES_REQUESTED** [review]
This an awesome chapter, well done. I have a few minor suggestions - then we can merge it. Thanks

File: docs/architecture/dapr-for-net-developers/secrects-buildingblock.md (hunk 1 above)
Comment (EdwinVW) [reply docs/architecture/dapr-for-net-developers/secrects-buildingblock.md:166]: We will do this during a last pass over the entire book.

**Comment by EdwinVW** [comment]
I've accepted all suggestions. So this chapter can be merged.

**Review by IEvangelist — APPROVED** [review]
Looks good to me, we'll :shipit: - thank you 🙏

**Pull request merged**